### ISSUE TITLE
This evening option, prefferably with customizable time

### ISSUE THREAD

**Issue by bubblefoil** [bwinton/SnoozeTabs#456]
I use my laptop both at work and home. Often I open a tab which I would like to get back to after work. Which is in the evening. Which sounds like "later today", although this option puts the tab off by just a few hours and I have to re-snooze the tabs repeatedly throughout the day.
I agree I could use the calendar but I am not a fan of this tiny mouse-wheel-scrollable combo, it is just too user unfriendly. Evening hours are too far down, scrolling with a wheel feels imprecise. I would prefer either a radial menu or an option to start dragging over the number to change it like with a slider.

**Comment by unhammer** [comment]
"This evening" should be a setting whenever it's between 9-5, and "next morning" when it's outside 9-5 :-)

**Comment by demoklion** [comment]
The UI of choosing time is unusable. Why not make a more standard one, where one can:

- write time in (eg. 9, 9pm, 18, 16:52, 1652)
- choose simply _later_, ie. if it's workday, show it the next morning, if it's weekend, on monday
- customisable other options, like eg. snoozing notifications on Android, snoozing e-mails on Gmail (RIP Inbox)

**Comment by xuiqzy** [comment]
Letting the user create custom snooze times would be a solution for a lot of requests here. This https://addons.mozilla.org/en-US/firefox/addon/snooze-tabby/?src=search unfortunately not open source addon for snoozing allows this for example. But having a fixed time and not only an offset would be really handy.
Should I rather create a new issue for this proposal?

**Comment by maxsu** [comment]
I'd like to add this  - can any maintainers get me up to speed on the process here?

Looks like we'd have to edit:

`src/icons`
- Add new icon

Eg, an evening version of the "Later Today" icon:

![image](https://user-images.githubusercontent.com/25123/64074374-29fe4580-cc5f-11e9-874e-5ad995e81ddb.png)



[src/lib/time.js](https://github.com/bwinton/SnoozeTabs/blob/870727f3e0279b14cb496b0546a595e0db81ea31/src/lib/times.js)
- Pick an id and locale message name (eg, `evening` and `timeEvening`)
- Extend the times const on lines 10 - 18
- Extend the case statement on lines 27 - 56

This is where a configurability requirement could begin to make things tricky. For now I'd say lets leave it out.

[src/popup/snooze.html](https://github.com/bwinton/SnoozeTabs/blob/870727f3e0279b14cb496b0546a595e0db81ea31/src/popup/snooze.html)
 - Prettyprint html
 - Add a `This Evening` item

Likewise, the popup would need significant work to add any kind of configurability of individual items.

locales/**/snoozetabs.properties
- Add a `timeEvening` message
- Localize message

I don't know anything about localizing at this time. Note however that if we allow users to configure their own items, we can reduce localization work.

**Comment by maxsu** [comment]
@xuiqzy adding configurable times sounds like a good goal. Want to collaborate on that? 

Questions:
1. How do users specify time and date offsets?
    - Fixed time of day + n days in the future (easy)
    - Time offset (+hours, minutes days)         (easy)
    - Weekday names                                      (medium, due to localization)
    - Custom rules                                           (tricky af)
2. Can users delete existing rules?
3. Should users be able to reset rules?
4. Should users be able to export or sync rules?
 
@demoklion How would you feel about an analog time picker with ability to type in the time? Wanna collaborate with me here?

![image](https://user-images.githubusercontent.com/25123/64074471-831aa900-cc60-11e9-925c-158724c858f8.png)

Also, these do seem like existing issues.
- Timepicker issue: #374
- Custom snooze settings: #270 

Feel free to drop a comment here, or mention me on one of those issues!

**Comment by demoklion** [comment]
@maxsu hey, I think the baseline for this could be how gmail on android does this now (minus the swiping interactions and ML smarts in the background). 
Using standard (or at least common) material components is a good start. They aren't perfect (material is mobile-first), but they're certaintly better than what we have now.
check eg. [this component in material ui](https://material-ui.com/components/pickers/#date-time-pickers)

Custom snooze settings are not that important to me, what's in the plugin now works fine. If choosing custom date/time is easier to do, that solves my issue.

**Comment by bubblefoil** [comment]
I might actually be able to help with the inner parts. I have written some tampernonkey scripts and would like to  learn the process of writing an extension. I am a UI noob, though.

**Comment by maxsu** [comment]
I've pushed a [commit](https://github.com/maxsu/SnoozeTabs/commit/d0559d868bf8571b4cb654172f7e8b768c317189) to a temporary [feature branch](https://github.com/maxsu/SnoozeTabs/tree/feature/thisEvening).

I don't think this will necessarily become a feature of the app, but it's a useful entry point into the code (and of course all are welcome to use my branch or pull the feature). Please review the code - it's very simple so far. 

![image](https://user-images.githubusercontent.com/25123/64345145-a321d380-cfa4-11e9-9daa-f5d7fea5a233.png)


Things to note:
1. I manually adjusted the height of the app's CSS to accommodate the new entry. Otherwise the app looks like this:
  ![image](https://user-images.githubusercontent.com/25123/64343790-06f6cd00-cfa2-11e9-96a6-4e954d257204.png)
  This may need to become automatic if users can add or delete entries. I am somewhat remiss in that I lack the CSS-fu to do this! Help welcome!

2. I didn't make an effort to localize the additions (even in an automated way). We may need to look into whether pontoon localization is still available, or find another open source localization service. I'd hate to scale back on localization.


To test the code, do:

1a. Clone my project at the feature branch:

> git clone -b feature/thisEvening https://github.com/maxsu/SnoozeTabs.git

OR..

1b. Checkout my project as a [remote](https://git-scm.com/book/en/v2/Git-Basics-Working-with-Remotes).

> cd src/SnoozeTabs
> git remote add maxsu https://github.com/maxsu/SnoozeTabs.git
> git checkout maxsu/feature/thisEvening

2. Build the project in [production mode](https://github.com/maxsu/SnoozeTabs/blob/master/package.json#L117)
> npm install
> npm run package

3. Install as a temporary plugin in Firefox

- Go to `Menu > Add-Ons > Tools for all addons > Load Temporary Addon`
- Load `SnoozeTabs/dist/manifest.json`

Enjoy!

**Comment by bubblefoil** [comment]
Why do I need python to build it?

**Comment by bwinton** [comment]
I don't believe you do? What error are you getting that suggests it's missing Python?

**Comment by bubblefoil** [comment]
The error said it could not find python.exe. So I added it to paths and the error changed a little, it was still Python related, though.
I'm afraid I cannot give you more info, my laptop died recently and I did not have a change to try it again since then.

**Comment by smartfonreddit** [comment]
As a user of this addon, would I need to download the updated xpi from elsewhere, or will the official one on AMO get updated in future?

**Comment by maxsu** [comment]
You have two options currently, both require building my branch of the repo (see my post above).

1. Run it as a temporary addon (`about:debugging` > Load Temporary Addon)
2. Load the xpi into Firefox developer edition (`about:config`  > set `xpinstall.signatures.required` to `false`)

Let me know if you get it working, or have trouble. Please post console logs in the latter case.

If you test it and feel the feature merits inclusion into AMO, I'll create a PR.

**Comment by bubblefoil** [comment]
Hi, I'm back with a replacement laptop and another attempt to build the forked version.
I got the same error, here is the console output and npm log:
https://gist.github.com/bubblefoil/f43633e8ff166879661fb0747f595418
https://gist.github.com/bubblefoil/4956939ecef098c480c954de697238c6

**Comment by maxsu** [comment]
Thanks for posting logs. I see what's happening - you are working in windows without protective WSL armor*! You are quite brave!

I can replicate your error. Here's what I did to fix it:

1. Install npm's `windows-build-tools`, which handles `node-gyp`'s python dependency. (see [this guide](https://spin.atomicobject.com/2019/03/27/node-gyp-windows/))

    From admin powershell, do:
    >npm install --global production windows-build-tools --vs2015
2. Open `package.json` and update `node-sass` from `3.13.0` to `4.12`.
    - Old version of `node-sass` was not compatible with Windows
3. Run `npm install`.
    - You can ignore the peer requirement and optional dependency warnings.
4. Run `npm run package` using git bash, if you have it available.
    - If you run this in `cmd`, you'll see a `mv` error; despite this error, the web extension will build and appear as a `.zip` file in the `./web-ext-artifacts` directory.
    - In either case, the raw extension's `manifest.json` will be in `./dist`

*) On a side note, the extension is definitely packaged to be built on linux, and in general I've had much better luck working with `node`, `python`, etc in microsoft's surprisingly decent WSL environment. You might want to consider moving to that - See instructions for [WSL](https://docs.microsoft.com/en-us/windows/wsl/install-win10) + [windows insider](https://insider.windows.com/en-us/how-to-pc/) + the [Ubuntu 18.04 WSL image](https://wiki.ubuntu.com/WSL). To get started with node, see my notes [here](https://gist.github.com/maxsu/2158e37ccb4cbcea6012ab602c1a1d19).

Good luck, and let me know if worked!
